src/lib/snarky_js_bindings/lib/snarky_js_bindings_lib.ml

```diff
@@ -0,0 +1,2673 @@
+module Backend = Kimchi_backend.Pasta.Vesta_based_plonk
+module Other_backend = Kimchi_backend.Pasta.Pallas_based_plonk
+module Impl = Pickles.Impls.Step
+module Other_impl = Pickles.Impls.Wrap
+module Challenge = Limb_vector.Challenge.Make (Impl)
+module Sc =
+  Pickles.Scalar_challenge.Make (Impl) (Pickles.Step_main_inputs.Inner_curve)
+    (Challenge)
+    (Pickles.Endo.Step_inner_curve)
+module Js = Js_of_ocaml.Js
+
+let console_log_string s = Js_of_ocaml.Firebug.console##log (Js.string s)
+
+let console_log s = Js_of_ocaml.Firebug.console##log s
+
+let raise_error s =
+  let s = Js.string s in
+  Js.raise_js_error (new%js Js.error_constr s)
+
+let raise_errorf fmt = Core_kernel.ksprintf raise_error fmt
+
+class type field_class =
+  object
+    method value : Impl.Field.t Js.prop
+
+    method toString : Js.js_string Js.t Js.meth
+
+    method toJSON : < .. > Js.t Js.meth
+
+    method toFields : field_class Js.t Js.js_array Js.t Js.meth
+  end
+
+and bool_class =
+  object
+    method value : Impl.Boolean.var Js.prop
+
+    method toBoolean : bool Js.t Js.meth
+
+    method toField : field_class Js.t Js.meth
+
+    method toJSON : < .. > Js.t Js.meth
+
+    method toFields : field_class Js.t Js.js_array Js.t Js.meth
+  end
+
+module As_field = struct
+  (* number | string | boolean | field_class | cvar *)
+  type t
+
+  let of_field (x : Impl.Field.t) : t = Obj.magic x
+
+  let of_field_obj (x : field_class Js.t) : t = Obj.magic x
+
+  let value (value : t) : Impl.Field.t =
+    match Js.to_string (Js.typeof (Obj.magic value)) with
+    | "number" ->
+        let value = Js.float_of_number (Obj.magic value) in
+        if Float.is_integer value then
+          let value = Float.to_int value in
+          if value >= 0 then Impl.Field.of_int value
+          else Impl.Field.negate (Impl.Field.of_int (-value))
+        else raise_error "Cannot convert a float to a field element"
+    | "boolean" ->
+        let value = Js.to_bool (Obj.magic value) in
+        if value then Impl.Field.one else Impl.Field.zero
+    | "string" -> (
+        let value : Js.js_string Js.t = Obj.magic value in
+        let s = Js.to_string value in
+        try
+          Impl.Field.constant
+            ( if
+              String.length s >= 2
+              && Char.equal s.[0] '0'
+              && Char.equal (Char.lowercase_ascii s.[1]) 'x'
+            then Kimchi_pasta.Pasta.Fp.(of_bigint (Bigint.of_hex_string s))
+            else Impl.Field.Constant.of_string s )
+        with Failure e -> raise_error e )
+    | "object" ->
+        let is_array = Js.to_bool (Js.Unsafe.global ##. Array##isArray value) in
+        if is_array then
+          (* Cvar case *)
+          (* TODO: Check this works *)
```

Comment:
I find this case worrying; won't we convert any JS array blindly to an OCaml object? (This can be a TODO github issue, doesn't need resolving here.)

Reply:
Tracked here: https://github.com/MinaProtocol/mina/issues/10554